### ISSUE TITLE
Excessive CPU & Memory usage in long running K8S pods

### ISSUE THREAD

**Issue by alexhollis** [neo4j/neo4j#12953]
_Reporting an issue my colleague posted about in community forums here_: [post link]( https://community.neo4j.com/t5/neo4j-graph-platform/neo4j-k8s-pods-start-slowing-down-after-a-while/m-p/61078#M36154 )

### **Post details:** 

Hello, our team is currently doing an implementation using Neo4j and we are facing some issues.

### **Short description**

Our structure is relatively simple:

- the nodes have only two properties of interest: id(int) and seed(bool) and both are Indexed

- there's only one relation type between nodes (CREATES).

What we are trying to achieve is getting the shortest path like so:
```sql
CALL apoc.cypher.runTimeboxed("MATCH p = ShortestPath((t:Entity {id: toInteger('1')})<-[:CREATES*]-(s:Entity{seed: true}))
                WITH p
                LIMIT 10
                WITH collect(p) as paths
                CALL apoc.convert.toTree(paths)
                YIELD value
                RETURN value;", {}, 15000)
```

We have 3 - 10 services that call the above in total roughly 10-20 times per second.

The neo4j pod currently has 8 cores assigned and 32gbs of memory.

### **The problem we are facing:**

After pods are running for a while, the CPU starts getting maxed out and everything slows down and we start seeing almost all transactions coming out as "execution expired". After forcing a restart on Neo4j, everything goes back to normal, and CPU usage drops from 8 cores maxed out to ~2 cores.

### **What we have tried:**

When transactions start to fail we pinpoint the group of pods with the issue, grab the queries that are failing, connect directly to the pod, and execute the failing queries manually. The failing queries when ran manually take around 30 - 100ms to finish.

We also run a PROFILE against the queries and the Memory consumption is around 300kbs, most of the time there are no DB scans and when those exist are on the single digits.

Has anyone faced the above or has any pointers to what might be happening?

_Edit: So far the biggest Neo4j DB we have is 120k Nodes and 4 Million relationships._

_However, we have seen the problem happening also with 75k Nodes and 4 Million relationships._

_Our graph depth is usually focused around 5 and doesn't usually go past 10 but data is dynamic that can change. We are expecting an upper limit of 20._

_Edit2: we have now seen it start failing with 3k entities and 22k relationships._

**Comment by ashantiel** [comment]
@alexhollis I've landed here looking for a solution to our problem, which seems to be pretty similar...
We have around 100k nodes and 500k relations, nodes has different properties (mostly 2 props, some small sets has up to 8 properties) and there is always one relation type. We are doing approx. 30 calls of small queries for different results per request. Our solution has 2GB memory on which 1.5GB is for neo4j.
After a week on k8s our dbs start to slow down pretty much, and the only solution is to restart neo4j (neo4j restart is enough).
In some rare occurences, pod can take most of the CPU available on k8s, but that's not the rule in our case.

Seems like kind of memory leak for a scenario where you have a lot of small, very quick queries returning small sets few times per second.

**Comment by rlch** [comment]
@alexhollis @ashantiel 

Have either of you resolved this issue? Experiencing the same symptoms on K8s with ~40k nodes and 130k relationships. All it takes is a simple stress-test of ~500 queries/min to buckle the entire database; with restarting being the only solution to bringing it back online.

**Comment by ashantiel** [comment]
@rlch nope, in my case we created a scheduled job that connects to the pod and restarts neo4j. I'm also waiting for memgraph project as stable enough to try it on production instead of this one, as performance in this scenario (lot of small queries) is not the top prio of this project :)

**Comment by MishaDemianenko** [comment]
Hi, @alexhollis described behavior does not look normal. 
To be able to suggest something more specific and understand what the system is doing we would need a couple of CPU profiles during the "abnormal" times.
The easiest way how to collect those profiles would be using JFR recording or using async profiler: https://github.com/jvm-profiling-tools/async-profiler
